src/neptune/new/handler.py

```diff
@@ -518,10 +627,53 @@ def track_files(self, path: str, destination: str = None, wait: bool = False) ->
             attr = self._container.get_attribute(self._path)
             if not attr:
                 attr = Artifact(self._container, parse_path(self._path))
-
-            self._container.set_attribute(self._path, attr)
+                self._container.set_attribute(self._path, attr)
 
             attr.track_files(path=path, destination=destination, wait=wait)
 
     def __delitem__(self, path) -> None:
         self.pop(path)
+
+
+class ExtendUtils:
+    @staticmethod
+    def validate_and_transform_to_extend_format(value):
+        """Preserve nested structure created by `Namespaces` and `dict_like` objects,
+        but replace all other values with single-element lists,
+        so work can be delegated to `extend` method."""
+        if isinstance(value, Namespace) or is_dict_like(value):
+            return {k: ExtendUtils.validate_and_transform_to_extend_format(v) for k, v in value.items()}
+        elif is_collection(value):
+            raise NeptuneUserApiInputException(
```

Comment:
Was custom exception a requirement from spec? Because I think `ValueError` is standard python exception for such cases and I think it allows you to include some custom message.

Reply:
I think it was discussed between PK and Mycek and custom exception here looks reasonable, we're changing the API and want to explicitly inform user that he uses new endpoint the wrong way.

Can you confirm @Herudaio